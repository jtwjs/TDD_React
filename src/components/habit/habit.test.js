import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";

import Habit from "./habit";

describe("habit", () => {
  const habit = {
    name: "taewoong",
    count: 28,
  };

  it("renders", () => {
    const component = renderer.create(
      <Habit
        habit={habit}
        onIncrement={jest.fn()}
        onDecrement={jest.fn()}
        onDelete={jest.fn()}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe("habit crud buttonn", () => {
    let onIncrement, onDecrement, onDelete, habitComponent;

    beforeEach(() => {
      onIncrement = jest.fn();
      onDecrement = jest.fn();
      onDelete = jest.fn();
      habitComponent = (
        <Habit
          habit={habit}
          onIncrement={onIncrement}
          onDecrement={onDecrement}
          onDelete={onDelete}
        />
      );
      render(habitComponent);
    });

    describe("increase", () => {
      let incrementBtn;
      beforeEach(() => {
        incrementBtn = screen.getByRole("button", { name: "increase" });
      });
      it("calls onIncreament when clicking 'increment' button", () => {
        userEvent.click(incrementBtn);

        expect(onIncrement).toHaveBeenCalledTimes(1);
      });
    });

    describe("decrease", () => {
      let decrementBtn;
      beforeEach(() => {
        decrementBtn = screen.getByRole("button", { name: "decrease" });
      });

      it("calls onDecreament when clicking 'decrement' button", () => {
        userEvent.click(decrementBtn);

        expect(onDecrement).toHaveBeenCalledTimes(1);
      });
    });

    describe("delete", () => {
      let deleteBtn;
      beforeEach(() => {
        deleteBtn = screen.getByRole("button", { name: "delete" });
      });
      it("calls onDelete when clicking 'decrement' button", () => {
        userEvent.click(deleteBtn);

        expect(onDelete).toHaveBeenCalledTimes(1);
      });
    });
  });
});
