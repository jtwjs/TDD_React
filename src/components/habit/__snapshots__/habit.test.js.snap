// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`habit renders 1`] = `
<li
  className="habit"
>
  <span
    className="habit-name"
    data-testid="name"
  >
    taewoong
  </span>
  <span
    className="habit-count"
    data-testid="count"
  >
    28
  </span>
  <button
    aria-label="increase"
    className="habit-button habit-increase"
    onClick={[Function]}
  >
    <i
      className="fas fa-plus-square"
    />
  </button>
  <button
    aria-label="decrease"
    className="habit-button habit-decrease"
    onClick={[Function]}
  >
    <i
      className="fas fa-minus-square"
    />
  </button>
  <button
    aria-label="delete"
    className="habit-button habit-delete"
    onClick={[Function]}
  >
    <i
      className="fas fa-trash"
    />
  </button>
</li>
`;
